execution/gethexec/caff_node.go

```diff
@@ -0,0 +1,259 @@
+package gethexec
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	espressoClient "github.com/EspressoSystems/espresso-sequencer-go/client"
+
+	"github.com/ethereum/go-ethereum/arbitrum_types"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/offchainlabs/nitro/arbos"
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+var (
+	retryTime = time.Second * 1
+)
+
+/*
+Caff Node creates blocks with finalized hotshot transactions
+*/
+type CaffNode struct {
+	stopwaiter.StopWaiter
+
+	config                  SequencerConfigFetcher
+	namespace               uint64
+	executionEngine         *ExecutionEngine
+	espressoClient          *espressoClient.Client
+	nextHotshotBlockNum     uint64
```

Comment:
If the node is down, should users check and set this number before restartting?

Reply:
Its not necessary but they can do it, if they want. 

Think of it this way, if `nextHotshotBlockNum` is not changed. The `messageWithMetadata` queue will then accumulate message which might have already been added to the database. But the `createBlock` function removes any such duplicates so we should be good.

Reply:
It does the removing here:
```
	// Check for duplicates and remove them
	if messageWithMetadataPos <= currentPos {
		log.Error("message has already been processed, removing duplicate",
			"messageWithMetadataPos", messageWithMetadataPos, "currentMessageCount", currentPos)
		n.messagesWithMetadata = n.messagesWithMetadata[1:]
		n.messagesWithMetadataPos = n.messagesWithMetadataPos[1:]
		return false
	}
```